src/darwin/Framework/CHIP/CHIPDeviceController.mm

```diff
@@ -93,27 +91,78 @@ - (instancetype)init
         if ([self checkForInitError:(_persistentStorageDelegateBridge != nullptr) logMsg:kErrorPersistentStorageInit]) {
             return nil;
         }
+    }
+    return self;
+}
 
-        [self getControllerNodeId];
-        errorCode = _cppCommissioner->Init(_localDeviceId, _persistentStorageDelegateBridge, _pairingDelegateBridge);
-        if ([self checkForInitError:(CHIP_NO_ERROR == errorCode) logMsg:kErrorCommissionerInit]) {
-            return nil;
+- (BOOL)isRunning
+{
+    __block BOOL commissionerInitialized;
+    dispatch_sync(_chipWorkQueue, ^{
+        commissionerInitialized = [self _isRunning];
+    });
+    return commissionerInitialized;
+}
+
+- (BOOL)_isRunning
+{
+    return _cppCommissioner != nullptr;
+}
+
+- (BOOL)shutdown
+{
+    dispatch_sync(_chipWorkQueue, ^{
+        if (_cppCommissioner) {
+            CHIP_LOG_DEBUG("%@", kInfoStackShutdown);
+            _cppCommissioner->Shutdown();
+            delete _cppCommissioner;
+            _cppCommissioner = nullptr;
         }
+    });
 
-        _chipSelectQueue = dispatch_queue_create(CHIP_SELECT_QUEUE, DISPATCH_QUEUE_SERIAL);
-        if ([self checkForInitError:(_chipSelectQueue != nil) logMsg:kErrorNetworkDispatchQueueInit]) {
-            return nil;
+    return YES;
+}
+
+- (BOOL)startup:(id<CHIPPersistentStorageDelegate>)storageDelegate queue:(nonnull dispatch_queue_t)queue
+{
+    __block BOOL commissionerInitialized = NO;
+    dispatch_sync(_chipWorkQueue, ^{
+        if ([self _isRunning]) {
+            commissionerInitialized = YES;
+            return;
+        }
+
+        CHIP_ERROR errorCode = CHIP_ERROR_INCORRECT_STATE;
+
+        _persistentStorageDelegateBridge->setFrameworkDelegate(storageDelegate, queue);
+        // initialize NodeID if needed
+        [self _getControllerNodeId];
+
+        _cppCommissioner = new chip::Controller::DeviceCommissioner();
+        if (_cppCommissioner != nullptr) {
+            errorCode = _cppCommissioner->Init(_localDeviceId, _persistentStorageDelegateBridge, _pairingDelegateBridge);
+        }
+        if ([self checkForStartError:(CHIP_NO_ERROR == errorCode) logMsg:kErrorCommissionerInit]) {
+            return;
         }
 
         // Start the IO pump
-        dispatch_async(_chipSelectQueue, ^() {
-            self.cppCommissioner->ServiceEvents();
-        });
-    }
-    return self;
+        self.cppCommissioner->ServiceEvents();
```

Comment:
Why `self.cppCommissioner` vs `_cppCommissioner`?  Is there a semantic difference?

Reply:
There is. `_cppCommissioner` refers to the member variable/property directly whereas `self.cppCommissioner` accesses it via the auto-generated "getter". I could use either here because we're not overriding the getter.